crates/core/transaction/src/is_action.rs

```diff
@@ -367,9 +367,65 @@ impl IsAction for Swap {
                         .cloned(),
                 }
             }
-            None => SwapView::Opaque {
-                swap: self.to_owned(),
-            },
+            None => {
+                // If we can find a matching BSOD in the TxP, we can use it to compute the output notes
+                // for the swap.
+                let bsod = txp
+                    .batch_swap_output_data
+                    .iter()
+                    // This finds the first matching one; there should only be one
+                    // per trading pair per block and we trust the TxP provider not to lie about it.
+                    .find(|bsod| bsod.trading_pair == self.body.trading_pair);
+
+                // We can get the denom metadata whether we get a BSOD or not
+
+                let denom_1 = txp.denoms.get(&self.body.trading_pair.asset_1()).cloned();
+                let denom_2 = txp.denoms.get(&self.body.trading_pair.asset_2()).cloned();
+
+                match bsod {
+                    None => {
+                        // If we can't find a matching BSOD, we can't compute the output notes
+                        // for the swap.
+                        // TODO: is defaulting to 'None' the right behavior here, or is not matching on a BSOD as above an error?
```

Comment:
this is a question i do not know the answer to!

Reply:
Not finding a BSOD in the TxP means that whoever generated the transaction perspective didn't include that data, either because they didn't have it (eg the transaction was created but never confirmed by the chain and executed) or because they didn't support it in their implementation.

We should handle that by graceful degradation, just not rendering the BSOD-requiring data if we don't have it.

Reply:
to prompt discussion: we end up on this branch if the swap ciphertext could not be decrypted with the perspective's payload key associated with this swap's commitment, _and_ no batch swap output data in the perspective matched this trading pair.

if so, does it make sense to alternatively return an error (_this function would return an `anyhow::Result<ActionView>`_), return nothing at all (_this function would return an `Option<ActionView>`_) or does filling `None` into these fields as-is make sense?

Reply:
(pardon me, didn't see the new message when drafting this.)

Reply:
pushed 5d3533f to remove this `TODO` comment.